### PR TITLE
MAISTRA-809: Force a push if we discovery a pod with an existing conn…

### PR DESCRIPTION
**Pull request by knrc**
…ection

### CODE REVIEW

File: pilot/pkg/proxy/envoy/v2/eds.go [hunk 1]
```diff
@@ -521,6 +521,23 @@ func (s *DiscoveryServer) WorkloadUpdate(id string, labels map[string]string, _
...
+			// then do a full xDS push for this proxy;
+			// otherwise:
+			//   case 1: the workload has no sidecar proxy, no need xDS push at all.
+			//   case 2: the workload xDS connection has not been established,
+			//           also no need to trigger a full push here.
+			if connection.modelNode.IPAddresses[0] == id {
+				// There is a possibility that the pod comes up later than endpoint.
+				// So no endpoints add/update events after this, we should request
+				// full push immediately to speed up sidecar startup.
+				go s.startPush(versionInfo(), s.globalPushContext(), false, nil, id)
```
Comment (dgn) [inline pilot/pkg/proxy/envoy/v2/eds.go:538]: The upstream PR seems to be triggering a full push for the connection in this case, shouldn't we do that, too?

**Comment by dgn** [comment]
This looks promising, I'm currently trying to verify that this resolves the issue

File: pilot/pkg/proxy/envoy/v2/eds.go (hunk 1 above)
Comment (dgn) [reply pilot/pkg/proxy/envoy/v2/eds.go:538]: Going through the logs, it seems sufficient to do an incremental push. Missing clusters and listeners are pushed out immediately when a late pod event comes in, before a full push is triggered.

**Comment by dgn** [comment]
I finally managed to get this to reproduce reliably by building with `golang==1.11`. Interestingly, when building using `golang>=1.12`, these errors happen much more rarely - but they do still happen. This was very confusing, because adding a debug statement and then building with my local golang installation (`1.12`) would cause the error rate to drop considerably. Average time to reproduction on `golang==1.11` is ~4min for me, whereas with `golang==1.12` I could only ever reproduce once after 90mins

Update: Nevermind, in attempt 3 I now have flaky reproduction again

File: pilot/pkg/proxy/envoy/v2/eds.go (hunk 1 above)
Comment (dgn) [reply pilot/pkg/proxy/envoy/v2/eds.go:538]: My only request here is: if we stick with incremental push, we should adjust the comment above this line that still mentions a full push.

**Review by rcernich — APPROVED** [review]

**Review by dgn — APPROVED** [review]

File: pilot/pkg/proxy/envoy/v2/eds.go (hunk 1 above)
Comment (dgn) [reply pilot/pkg/proxy/envoy/v2/eds.go:538]: Apparently this is a misunderstanding on my part about incremental vs full push

**Review by brian-avery — APPROVED** [review]

**Comment by knrc** [comment]
Updated comments to refer to the related upstream issue and squashed

**Pull request merged**